eth/gasprice/locaFeeMarket/nativediscount.go

```diff
@@ -0,0 +1,318 @@
+	transferFunction = []byte{0xa9, 0x05, 0x9c, 0xbb}
+
+	// erc20's transferFrom: the first 4 bytes of the keccak256 hash of "transferFrom(address,address,uint256)"
+	transferFromFunction = []byte{0x23, 0xb8, 0x72, 0xdd}
+
+	// erc20's approve: the first 4 bytes of the keccak256 hash of "approve(address,uint256)"
+	approveFunction = []byte{0x09, 0x5e, 0xa7, 0xb3}
+
+	debugDiscountedGasPrice = new(big.Int).SetUint64(18_000_000_000)
+
+	errNotFunctionInvocation = errors.New("tx is not a function invocation")
+
+	//-------
+	btcLstContract = common.HexToAddress("0xTBD")
+
+	internalErc20Whitelist = map[common.Address]bool{
+		btcLstContract: true,
+	}
+)
+
+
+func AdjustGasPriceForEstimation(_networkGasPrice *hexutil.Big, _gasAmount *hexutil.Uint64, _value *hexutil.Big, to *common.Address, data []byte) *hexutil.Big {
+	if _networkGasPrice == nil || _gasAmount == nil || _value == nil {
+		return _networkGasPrice
+	}
+	networkGasPrice := (*big.Int)(_networkGasPrice)
+	gasAmount := uint64(*_gasAmount)
+	value := (*big.Int)(_value)
+	adjusted := AdjustGasPrice(gasAmount, value, to, data, networkGasPrice)
+	return (*hexutil.Big)(adjusted)
+}
+
+func AdjustGasPrice(gasAmount uint64, value *big.Int, to *common.Address, data []byte, networkGasPrice *big.Int) *big.Int {
+	// apply gas-price discount if tx is a 'native' transfer of either Core or whitelisted erc20
+	if isCoreTransferTx(gasAmount, value, len(data)) {
+		return discountCoreTransferTx(networkGasPrice)
+	}
+
+	if ftype := isErc20TransferTx(to, data); ftype != ftype_none {
+		return discountErc20TransferTx(networkGasPrice, ftype, gasAmount)
+	}
+
+	return networkGasPrice
+}
+
+
+func isCoreTransferTx(gasAmount uint64, value *big.Int, dataLen int) bool {
+	// requires (a) EOA destination (b) value > 0 (c) empty data indicating not a smart-contract
+	// invocation and (d) total gasAmount == 21k so to rule out transfer into a smart contract (and not EOA)
+	hasValue := value != nil && value.Sign() > 0
+	eoaToAddress := gasAmount == coreTransferGasAmount
+	isCoreTransfer := hasValue && eoaToAddress && dataLen == 0
+	return isCoreTransfer
+}
+
+func discountCoreTransferTx(networkGasPrice *big.Int) *big.Int {
+	// apply gas-price discount for Core transfer tx
+	if useDebugDiscountedGasPrice {
+		log.Debug("@lfm: debug-mode core transfer discount", "networkGasPrice", networkGasPrice, "debugGasPrice", debugDiscountedGasPrice)
+		return debugDiscountedGasPrice
+	}
+	discountedGasPrice := calcCoreTransferDiscountedGasPrice(networkGasPrice)
+	log.Debug("@lfm: core transfer discount", "networkGasPrice", networkGasPrice, "discounted-gas-price", discountedGasPrice)
+	return discountedGasPrice
+}
+
+func calcCoreTransferDiscountedGasPrice(networkGasPrice *big.Int)  *big.Int {
+	// calc discounted gas price for native Core transfer tx
+	return internalCalcDiscountedGasPrice(networkGasPrice)
+}
+
+func obsolete__internalCalcDiscountedGasPrice(networkGasPrice *big.Int)  *big.Int {
```

Comment:
`obsolete__internalCalcDiscountedGasPrice` has 0 references, what is the justification for not removing it? If it's just here for documentation purposes let's move it to a dedicated testing or documentation file. If this information is important enough to be colocated with the actual `internalCalcDiscountedGasPrice` function then we should condense the information and add it as inline comments to the `internalCalcDiscountedGasPrice` function.

Reply:
yep I left is for reviewers to diff. now removed